client/shaders/model_shader/opengl_fragment.glsl

```diff
@@ -0,0 +1,133 @@
+
+varying vec3 vNormal;
+varying vec3 vPosition;
+varying vec3 worldPosition;
+
+varying vec3 eyeVec;
+varying vec3 lightVec;
+
+bool normalTexturePresent = false;
+bool texTileableHorizontal = false;
+bool texTileableVertical = false;
+bool texSeamless = false;
+
+const float e = 2.718281828459;
+const float BS = 10.0;
+const float fogStart = FOG_START;
+const float fogShadingParameter = 1 / ( 1 - fogStart);
+
+void get_texture_flags()
+{
+	vec4 flags = texture2D(textureFlags, vec2(0.0, 0.0));
+	if (flags.r > 0.5) {
+		normalTexturePresent = true;
+	}
+	if (flags.g > 0.5) {
+		texTileableHorizontal = true;
+	}
+	if (flags.b > 0.5) {
+		texTileableVertical = true;
+	}
+	if (texTileableHorizontal && texTileableVertical) {
+		texSeamless = true;
+	}
+}
+
+float intensity(vec3 color)
+{
+	return (color.r + color.g + color.b) / 3.0;
+}
+
+float get_rgb_height(vec2 uv)
+{
+	if (texSeamless) {
+		return intensity(texture2D(baseTexture, uv).rgb);
+	} else {
+		return intensity(texture2D(baseTexture, clamp(uv, 0.0, 0.999)).rgb);
+	}
+}
+
+vec4 get_normal_map(vec2 uv)
+{
+	vec4 bump = texture2D(normalTexture, uv).rgba;
+	bump.xyz = normalize(bump.xyz * 2.0 - 1.0);
+	return bump;
+}
+
+void main(void)
+{
+	vec3 color;
+	vec4 bump;
+	vec2 uv = gl_TexCoord[0].st;
+	bool use_normalmap = false;
+	get_texture_flags();
+
+#if USE_NORMALMAPS == 1
+	if (normalTexturePresent) {
+		bump = get_normal_map(uv);
+		use_normalmap = true;
+	}
+#endif
+
+#if GENERATE_NORMALMAPS == 1
+	if (normalTexturePresent == false) {
+		float tl = get_rgb_height(vec2(uv.x - SAMPLE_STEP, uv.y + SAMPLE_STEP));
+		float t  = get_rgb_height(vec2(uv.x - SAMPLE_STEP, uv.y - SAMPLE_STEP));
+		float tr = get_rgb_height(vec2(uv.x + SAMPLE_STEP, uv.y + SAMPLE_STEP));
+		float r  = get_rgb_height(vec2(uv.x + SAMPLE_STEP, uv.y));
+		float br = get_rgb_height(vec2(uv.x + SAMPLE_STEP, uv.y - SAMPLE_STEP));
+		float b  = get_rgb_height(vec2(uv.x, uv.y - SAMPLE_STEP));
+		float bl = get_rgb_height(vec2(uv.x -SAMPLE_STEP, uv.y - SAMPLE_STEP));
+		float l  = get_rgb_height(vec2(uv.x - SAMPLE_STEP, uv.y));
+		float dX = (tr + 2.0 * r + br) - (tl + 2.0 * l + bl);
+		float dY = (bl + 2.0 * b + br) - (tl + 2.0 * t + tr);
+		bump = vec4(normalize(vec3 (dX, dY, NORMALMAPS_STRENGTH)), 1.0);
+		use_normalmap = true;
+	}
+#endif
+
+	vec4 base = texture2D(baseTexture, uv).rgba;
+
+#ifdef ENABLE_BUMPMAPPING
+	if (use_normalmap) {
+		vec3 L = normalize(lightVec);
+		vec3 E = normalize(eyeVec);
+		float specular = pow(clamp(dot(reflect(L, bump.xyz), E), 0.0, 1.0), 1.0);
+		float diffuse = dot(-E,bump.xyz);
+		color = (diffuse + 0.1 * specular) * base.rgb;
+	} else {
+		color = base.rgb;
+	}
+#else
+	color = base.rgb;
+#endif
+
+	vec4 col = vec4(color.rgb, base.a);
+	float diffuseIntensity = dot(vNormal, vec3(0.0, 1.0, 0.2));
+	vec3 diffuseColor = vec3(1.0) * clamp(diffuseIntensity, -0.3, 1.0);
+	vec3 lightColor = emissiveColor.rgb;
+
+	col.rgb *= lightColor * 0.8 + (lightColor + 0.1) * diffuseColor * 0.3;
```

Comment:
Directional ambient lighting could give better results: https://github.com/HybridDog/minetest/blob/19d7a6c8d7c5a287c6fedbefe4f7d9828aaa497c/client/shaders/nodes_shader/opengl_fragment.glsl#L85-L107

Diffuse lighting with a single light direction has the disadvantage that there are always at least three faces on the cube with the same light strength, unless you do it incorrectly (here you allow negative diffuseIntensity values).
Directional ambient light is a bit more realistic than constant ambient light combined with diffuse light with constant light direction because directional ambient light approximately incorporates indirect lighting, which is similar to Physically Based Rendering I think.

Reply:
I'll try it now. But it looks like a lot of operations only to make other faces to look brighter. Maybe with a few more dot products (i.e. light sources) we can achieve the same effect?

Reply:
![изображение](https://user-images.githubusercontent.com/12214069/74101447-eb907f00-4b4a-11ea-85eb-57632313aae5.png)

Here. With just one additional "light" source it doesn't look so flat on the sides. And it's almost as cheap as dirt:

```GLSL

	// Top "light"
	float diffuseIntensity = dot(vNormal, normalize(vec3(0.0, 1.0, 0.0)));
	// Front "light"
	diffuseIntensity += dot(vNormal, normalize(vec3(0.0, 0.2, 0.7))) * 0.5;

	// No more negative "light"
	diffuseIntensity = clamp(diffuseIntensity, 0.0, 1.0);

	col.rgb *= lightColor * 0.7 + (lightColor + 0.1) * diffuseIntensity * 0.3;
```

Reply:
> But it looks like a lot of operations only to make other faces to look brighter.

It's because that code supports bumpmapping, where the normal vectors from the normal map need to be rotated to the surface face rotation. (normalmaps are in tangent space)
Without bumpmapping support, the code should be fairly simple, and can be put in the vertex shader:
```glsl
// untested
float directional_ambient(vec3 normal)
{
	vec3 v = normal * normal;
	if (normal.y < 0)
		return dot(v, vec3(0.670820f, 0.447213f, 0.836660f));
	return dot(v, vec3(0.670820f, 1.000000f, 0.836660f));
}
```
Original code, which is used for the nodes: https://github.com/minetest/minetest/blob/b14aa1c84714a4800d214768ff7868a7bb76f7ae/src/client/mesh.cpp#L45-L63

Reply:
Ok, that model better follows overall shading model. Done.

Reply:
"Use shading that matches light distribution on world nodes"
Yes, this seems a good idea.